matlab/core/camera_warp.m

```diff
@@ -0,0 +1,107 @@
+function [ errors, J_warp ] = camera_warp( image_keyframe, depths_keyframe, image_current, T, intrinsics, plot )
+% projects points into world, transforms and projects back
+%
+% TODO: write some tests for this function! There are some critical things
+% with the inverted V-axis...
+%
+% INPUT:
+%
+% image_keyframe:
+%   H * W intensity image, this will be warped
+%
+% depths_keyframe:
+%   H * W depths of keyframe image
+%
+% image_current:
+%   H * W intensity image, this is the target
+%
+% T:
+%   [x y z alpha beta gamma]' transformation to apply
+%
+% intrinsics:
+%   camera intrinsics
+% 
+
+[H, W] = size(image_keyframe);
+
+calc_jacobian = nargout > 1;
+show_plots = nargin > 5;
+
+% check input parameters
+assert(all(size(depths_keyframe) == [H,W]), 'depth matrix must have same size as intensity image');
+assert(all(size(image_current)   == [H,W]), 'current image must have same size as keyframe image');
+assert(numel(T) == 6);
+assert(isa(intrinsics, 'CameraIntrinsics'));
+
+
+[X,Y] = meshgrid(1:W, 1:H);
+
+points_keyframe_camera = image_to_list(cat(3,X,Y));
+
+% attention: inverted V-Axis -> flip depths!
+points_world = camera_project_inverse(points_keyframe_camera, image_to_list(flipud(depths_keyframe)), intrinsics);
+
+if calc_jacobian
+    [points_current,        J_T] = camera_transform(points_world, T);
+    [points_current_camera, J_P] = camera_project(points_current, intrinsics);
+else
+    points_current        = camera_transform(points_world, T);
+    points_current_camera = camera_project(points_current, intrinsics);
+end
+
+
+
+% filter points outside of image range
+valid_points              = is_in_img_range(points_current_camera, image_current);
+points_current_camera     = points_current_camera(:, valid_points);
+
+if calc_jacobian
+    J_T = J_T(:,:,valid_points);
+    J_P = J_P(:,:,valid_points);
+    [intensities_current, J_I] = camera_intensity_sample(points_current_camera, image_current);
```

Comment:
Ich glaub die Jacobian sollte vom keyframe berechnet werden. Scheint ein typo in Sammy's paper eq. 11 zu sein. Im text spricht er von keyframe.